### DISCUSSION TITLE
How to define complex update input

### DISCUSSION THREAD

**Discussion by alveshelio** [trpc/trpc#2507] [Q&A] [answered]
Hi,

I'm quite new to tRPC, I have experience with GraphQL and I'm trying to understand how I can define the `input` in update mutations. I'm not sure how to proceed, the input could be quite complex based on the relations between tables.

I have this model
```prisma
enum ClientType {
  INDIVIDUAL
  COMPANY
}

enum Category {
  IRRIGATION_SYSTEMS
  DECKING_FENCING
  PLANTS
  PAVING
}

enum InvoiceStatus {
  PAID
  UNPAID
}

model Client {
  id                  String      @id @default(uuid())
  firstName           String
  lastName            String
  street              String
  postalCode          String
  city                String
  email               String      @unique
  type                ClientType
  createdAt           DateTime    @default(now())
  updatedAt           DateTime    @updatedAt
  phoneNumberMobile   String?
  phoneNumberLandLine String?
  invoices            Invoice[]
  quotations          Quotation[]

  @@index([email, firstName, lastName, phoneNumberMobile, phoneNumberLandLine])
}

model Stock {
  id       String   @id @default(uuid())
  name     String
  quantity Int
  price    Float
  category Category
}

model Invoice {
  id          String           @id @default(uuid())
  number      Int              @unique
  createdAt   DateTime         @default(now())
  updatedAt   DateTime         @updatedAt
  status      InvoiceStatus
  dueOn       DateTime
  clientId    String
  client      Client           @relation(fields: [clientId], references: [id], onDelete: NoAction)
  quotationId String?          @unique
  quotation   Quotation?       @relation(fields: [quotationId], references: [id], onDelete: NoAction)
  sections    InvoiceSection[]

  @@index([number, dueOn])
}

model Quotation {
  id         String   @id @default(uuid())
  number     Int      @unique
  createdAt  DateTime @default(now())
  updatedAt  DateTime @updatedAt
  validUntil DateTime

  clientId String
  client   Client             @relation(fields: [clientId], references: [id], onDelete: NoAction)
  invoice  Invoice?
  sections QuotationSection[]

  @@index([number, validUntil])
}

enum SectionUnit {
  HOURLY
  QUANTITY
}

model InvoiceSectionItem {
  id          String           @id @default(uuid())
  name        String
  price       String
  description String           @db.Text
  sectionId   String
  sections    InvoiceSection[]

  @@index([sectionId])
}

model InvoiceSection {
  id        String               @id @default(uuid())
  name      String
  category  Category
  items     InvoiceSectionItem[]
  unit      SectionUnit
  invoiceId String
  invoice   Invoice              @relation(fields: [invoiceId], references: [id], onDelete: NoAction)
}

model QuotationSectionItem {
  id          String             @id @default(uuid())
  name        String
  price       String
  description String             @db.Text
  sectionId   String
  sections    QuotationSection[]

  @@index([sectionId])
}

model QuotationSection {
  id          String                 @id @default(uuid())
  name        String
  category    Category
  items       QuotationSectionItem[]
  unit        SectionUnit
  quotationId String?
  quotation   Quotation?             @relation(fields: [quotationId], references: [id], onDelete: NoAction)
}
```

I have  a route to update an invoice. In the invoice I want to be able to add/remove/update sections and/or items to these sections. In GraphQL I could use GraphQL code generator to generate the type for the `add`, `remove`, `update` and in the client I could use this type to type the output of a function that would build the input for the mutation. 

Is there a good example of a more or less complex input for an update mutation or could you guys guide me on what would be the best approach to define this kind of input?

Thank you

**Comment by alveshelio** [comment]
Anyone?

**Reply by jlalmes** [reply]
Not entirely sure what you're specifically asking for. 

Here's an example of what your `"update"` tRPC procedure could look like:

```ts
.mutation('updateInvoice', {
  input: z.object({
    id: z.string(),
    status: z.nativeEnum(InvoiceStatus).optional(),
    dueOn: z.date().optional(),
    /* ... etc. */
  }),
  resolve: async ({ input }) => {
    return prisma.invoice.update({
      where: { id: input.id },
      data: {
        status: input.status,
        dueOn: input.dueOn,
        /* ... etc. */
      }
    })
  }
}

```

**Comment by alveshelio** [comment, marked as answer]
Hi @jlalmes ,

Thank you for getting back. I'm looking for updates that have relationships. In my example, I have an invoice that has sections that has items. I'm wondering if I need to fully define what kind of object that I'm expecting. This could become quite complex based on the type of relationships you can get.

In GraphQL I can create a function that would build the variables to the needed input. I could simply type the function return with the  appropriate type and the function would take care of building the input accordingly to the needed output.

What are the best practices for complex update inputs? Are we supposed to define the input exactly as the type it's expecting?

**Reply by jlalmes** [reply]
> Are we supposed to define the input exactly as the type it's expecting?

Yes.

A tool like [`zod-prisma`](https://github.com/CarterGrimmeisen/zod-prisma) might help you to codegen `zod` schemas for you models.

**Reply by alveshelio** [reply]
Hi @jlalmes Thank you very much.

**Reply by KATT** [reply]
I do a bunch of very complex inputs & using stuff like [`z.discriminatedUnion`](https://github.com/colinhacks/zod#discriminated-unions) :)

**Reply by alveshelio** [reply]
Hi @KATT,

Thank you very much for getting back. I actually find it quite tedious to build complex schemas. I which there was something a bit better that would simply build a schema out of a type. I took a look at `zod-prisma` but it doesn't go deep into all the nested inputs.

The only experience I have with GraphQL is with Hasura GraphQL Engine and it would do the heavy lifting for me, it would generate the update schema and so I could simply create a function that would then make sure to build the input for the update. I guess, that if I want to use tRPC, then I'd be the one who would need to create the appropriate schema.

**Reply by jlalmes** [reply]
I suppose you could submit a PR to `zod-prisma` such that it would traverse the relationship trees of your models.

Sadly, at this time, there's no great tools for using types as values (even at compile-time). I've played with [`ts-morph`](https://github.com/dsherret/ts-morph) before which is super powerful, but the DX is really clunky.

**Reply by u007** [reply]
zod-prisma looks dead now, how do we solve recursive input schema?
im trying to avoid this:
 WARN  Recursive reference detected at #/properties/orderBy/anyOf/1/items/properties/order/properties/site/properties/defaultRedisService/properties/owner/properties/organisation/properties/Site! Defaulting to any

or at least somewhere i can set the recusive level to 5 level max maybe?
